apps/prairielearn/src/pages/enroll/enroll.sql

```diff
@@ -44,23 +44,93 @@ ORDER BY
   ci.id DESC;
 
 -- BLOCK enroll
-INSERT INTO
-  enrollments AS e (user_id, course_instance_id)
+WITH
+  enrollment_limits AS (
+    SELECT
+      i.id AS institution_id,
+      i.yearly_enrollment_limit AS institution_yearly_enrollment_limit,
+      -- Course instance enrollment limit overrides institution course instance
+      -- enrollment limit to allow for higher limits on individual course instances.
+      COALESCE(
+        ci.enrollment_limit,
+        i.course_instance_enrollment_limit
+      ) AS course_instance_enrollment_limit
+    FROM
+      institutions AS i
+      JOIN pl_courses AS c ON (c.institution_id = i.id)
+      JOIN course_instances AS ci ON (ci.course_id = c.id)
+    WHERE
+      ci.id = $course_instance_id
+      -- Lock institution to prevent concurrent enrollments that would exceed
+      -- the enrollment limit.
+    FOR UPDATE OF
+      i
+  ),
+  course_instance_enrollments AS (
+    SELECT
+      COUNT(*) AS count
+    FROM
+      enrollments
+    WHERE
+      course_instance_id = $course_instance_id
+  ),
+  institution_enrollments AS (
+    SELECT
+      COUNT(*) AS count
+    FROM
+      enrollments AS e
+      JOIN course_instances AS ci ON (e.course_instance_id = ci.id)
+      JOIN pl_courses AS c ON (ci.course_id = c.id)
+      JOIN institutions AS i ON (i.id = c.institution_id)
+      JOIN enrollment_limits AS el ON (el.institution_id = i.id)
+    WHERE
+      e.created_at > now() - interval '1 year'
+  ),
+  limits_exceeded AS (
+    SELECT
+      COALESCE(
+        cie.count + 1 > el.course_instance_enrollment_limit,
+        FALSE
+      ) AS course_instance_limit_exceeded,
+      COALESCE(
+        ie.count + 1 > el.institution_yearly_enrollment_limit,
+        FALSE
+      ) AS institution_yearly_limit_exceeded
+    FROM
+      course_instance_enrollments AS cie
+      CROSS JOIN institution_enrollments AS ie
+      CROSS JOIN enrollment_limits AS el
+  ),
+  new_enrollment AS (
+    INSERT INTO
+      enrollments AS e (user_id, course_instance_id)
+    SELECT
+      u.user_id,
+      $course_instance_id
+    FROM
+      users AS u,
+      limits_exceeded
+    WHERE
+      u.user_id = $user_id
+      AND check_course_instance_access (
+        $course_instance_id,
+        u.uid,
+        u.institution_id,
+        $req_date
+      )
+      AND NOT course_instance_limit_exceeded
+      AND NOT institution_yearly_limit_exceeded
+    ON CONFLICT DO NOTHING
+    RETURNING
+      e.id
+  )
 SELECT
-  u.user_id,
-  $course_instance_id
+  (
+    course_instance_limit_exceeded
+    OR institution_yearly_limit_exceeded
+  ) AS enrollment_limit_exceeded
 FROM
-  users AS u
-WHERE
-  u.user_id = $user_id
-  AND check_course_instance_access (
-    $course_instance_id,
-    u.uid,
-    u.institution_id,
-    $req_date
-  )
-RETURNING
-  e.id;
+  limits_exceeded;
```

Comment:
Fwiw I've been moving towards doing this kind of operation in TS, wrapping it in a `runTransactionAsync`, starting with a `lockCourseInstance()`, then doing a query to get the various enrollment counts, checking limits in TS, then doing the actual INSERT. I've found this to be easier to read debug. This style is facilitated by using `selectAndAuthz...` functions (sorry to keep harping on about these, but they really are nice 😀) and passing whole objects around rather than row IDs.

Reply:
That's fair! I can move to doing that. I'm still in "minimize DB roundtrips" mindset 😅

Reply:
Me too! But I've been getting over it and the resulting code has been much easier to deal with (plus typechecking!).